TypeScript.YAML-tmLanguage

```diff
@@ -200,50 +200,78 @@ repository:
   type-annotation:
     name: meta.type.annotation.ts
     begin: ":"
-    end: (?=[,);=>}\]\[])|(?<=[a-z]|>)\s*(?=\{|$)
+    end: (?=[,);=>}\]])|(?<=[a-z]|>)\s*(?=\{|$)
     patterns:
       - include: '#type'
 
   type:
     name: meta.type.ts
     patterns:
+    - include: '#type-intrinsic'
     - include: '#type-name'
     - include: '#type-parameters'
-    - include: '#type-union'
     - include: '#object-type'
-    - include: '#function-type-parameters'
+    - include: '#tuple-type'
+    - include: '#paren-type-or-function-type-parameters'
     - include: '#function-type-return-type'
+    - include: '#type-operators'
+    - include: '#comment'
 
-  function-type-return-type:
-    name: meta.function.type.return.ts
-    begin: '=>'
-    end: (?=,|\))
+  type-intrinsic:
+    name: storage.type.ts
+    match: '\b(string|number|boolean|symbol|any|void)\b'
+
+  type-name:
+    name: entity.name.type.ts
+    match: '([a-zA-Z_$][\w$]*)'
+
+  type-parameters:
+    name: meta.type.parameters.ts
+    begin: '<'
+    beginCaptures:
+      '0': { name: keyword.operator.type.ts }
+    end: '(?=var|type|function|class|interface|[)}:;\]])|>'
+    endCaptures:
+      '0': { name: keyword.operator.type.ts }
     patterns:
     - include: '#type'
 
-  type-union:
-    name: meta.type.union.ts
-    begin: '(\|)'
+  type-operators:
+    name: keyword.operator.type.ts
+    match: '[.|,]'
+
+  tuple-type:
+    name: meta.type.tuple.ts
+    begin: '\['
     beginCaptures:
-      '1': { name: keyword.operator.ts }
-    end: '([a-zA-Z_$][.\w$]*)'
+      '0': { name: keyword.operator.type.ts }
+    end: '\]'
     endCaptures:
-      '1': { name: storage.type.ts }
+      '0': { name: keyword.operator.type.ts }
     patterns:
-    - include: '#comment'
-
-  type-name:
-    name: storage.type.ts
-    match: '[a-zA-Z_$][.\w$]*'
+    - include: '#type'
 
-  type-parameters:
-    name: meta.type.parameters.ts
-    begin: '<'
-    end: '(?=var|type|function|class|interface)|>'  
+  paren-type-or-function-type-parameters:
+    name: meta.type.paren.cover.ts
+    begin: '\('
+    end: '\)'
     patterns:
-    - include: '#comment'
+    - include: '#type-or-function-type-parameter'
     - include: '#type'
 
+  type-or-function-type-parameter:
+    name: meta.type.paren.cover.ts
+    match: '(\.\.\.)?\s*([a-zA-Z_$][\w$]*)\s*(\?)?\s*(:)?\s*(?=,|\)\s*=>)'
+    captures:
```

Comment:
The first capture group captures the literal `...` which we use for rest arguments, which is why it is escaped as `\.`.